### PR TITLE
[NVIDIA GPU] Flatten call graph after collective pipeliner pass

### PR DESCRIPTION
**Pull request by Tixxx**
After moving the collective pipeliner pass after post layout optimization passes, we need to call flatten call graph after pipeliners have run so call sites are unique. We didnt need to do this before because layout assignment(which is after previous location of collective pipeline) runs flattenCallGraph.

### CODE REVIEW

**Review by frgossen — CHANGES_REQUESTED** [review]
Can you add a test case to help understand why this breaks w/o extra flattening?

**Comment by Tixxx** [comment]
> Why do we have two paths for pipelining the collectives anyway?
> 
> It seems to me this post-layout path was introduced on ["let's see if it breaks anything"](https://github.com/openxla/xla/pull/12866#issuecomment-2138240159) basis. Now, that we see there are some flaws to this, in my mind it makes sense to disable this by default.
> 
> Also please add a test for the case you provide a fix.

I think the plan was to only keep the pipeliner passes after all the post-layout optimizations have run. Pipeliners have a few side effects on some of the gemm fusions we have in post-layout optimizations, so we'd like to run them after target-specific fusions have happened which will result in simpler graphs. This is a bug that we didn't have full test coverage before, since there's no fundamental flaw of moving the pipeline, I think it'd be good to keep the original plan? We are still tracking the removal of the original pipeliners on our side.
I'm working on minimizing the model to have an e2e test for this case.

**Comment by golechwierowicz** [comment]
> This is a bug that we didn't have full test coverage before, since there's no fundamental flaw of moving the pipeline, I think it'd be good to keep the original plan?

Fundamentally the plan sounds good, however I've seen one instance of RS not being pipelined recently (I will share the HLO soon-ish) so this is why I have some doubts, mostly how freely we can move passes around given the robustness of collective pipeliner.

I think ultimately we don't want to see additional patches to make some specific cases work in collective pipeliner as it is already quite complex.

**Comment by Tixxx** [comment]
> > This is a bug that we didn't have full test coverage before, since there's no fundamental flaw of moving the pipeline, I think it'd be good to keep the original plan?
> 
> Fundamentally the plan sounds good, however I've seen one instance of RS not being pipelined recently (I will share the HLO soon-ish) so this is why I have some doubts, mostly how freely we can move passes around given the robustness of collective pipeliner.
> 
> I think ultimately we don't want to see additional patches to make some specific cases work in collective pipeliner as it is already quite complex.

I understand the concern. We are kind of choosing the less of 2 evils.
If we keep the pipeliner where it is, then we'd need a lot more logic in gemm rewriter to deal with pipelined QDQ nodes(which potentially could cause more bugs);
If we move pipeliner passes, we need to fix bugs like this before it stabilizes.
So given the concerns from your side, would keeping the 2 callsites of collective pipeliner and defaulting to the original callsite be a better option going forward?

We have noticed some issues with RS on our side too, some investigations suggest that we likely dont need to change pipeliner. Some transformations in gemm rewriter have side effects which can be avoided. Once you have the hlo, we can confirm if it's the same root cause.

**Comment by Tixxx** [comment]
@frgossen @golechwierowicz  added an e2e test that repros the crash.

File: xla/tests/collective_ops_e2e_test.cc [hunk 1]
```diff
@@ -981,5 +981,98 @@ ENTRY entry {
...
+TEST_F(CollectiveOpsTestE2E, PostLayoutCollectivePipelinerShouldFlattenCallGraph) {
+// The allgather in the loop has a nested while loop as its operand,
+// when the pipelining happens, the nested while loop will be peeled outside.
+// However, when a while is cloned, its callsites are still preserved which will
+// error out in alias analysis. When the graph is flattened, the error should not
+// happen.
+  absl::string_view kModuleReplicatedStr = R"(
+HloModule module
+
+while_cond {
```
Comment (frgossen) [inline xla/tests/collective_ops_e2e_test.cc:994]: Is this a minimal test case reproducing teh issue?

**Review by frgossen — CHANGES_REQUESTED** [review]

File: xla/tests/collective_ops_e2e_test.cc (hunk 1 above)
Comment (Tixxx) [reply xla/tests/collective_ops_e2e_test.cc:994]: Yes, we need the following conditions to repro the crash:
1. nested while loop
2. inner while loop being an operand of a pipelined collective which will be peeled out of the outer loop
The test just contains the hlo to meet the above conditions.

**Comment by Tixxx** [comment]
> > So given the concerns from your side, would keeping the 2 callsites of collective pipeliner and defaulting to the original callsite be a better option going forward?
> 
> I'd say we need to align on this a little bit more, pipelines are kind of a mess right now and it would be good to not add additional complexity there. Maybe I will write a small doc on how I would see the future of pipelines related to horizontal scaling (pipeliners, collective rewriters etc.) so we can get a consensus.
> 
> For the time being let's apply your suggestion of defaulting to the old callsite.
> 
> Thanks for your input!

That sounds good. IMO collective optimizations ideally should happen after "local" optimizations i.e gemm fusions or kernel fusions, but that also means the current collective passes would need to handle additional ops like bitcasts or converts from float normalizations which adds more complexity. Having a doc to brainstorm sounds great to me.

**Comment by golechwierowicz** [comment]
> > For the time being let's apply your suggestion of defaulting to the old callsite.
> 
> @golechwierowicz Can we put this in a separate pr?

Ok, let's put it in a separate PR.

**Review by golechwierowicz — APPROVED** [review]

**Comment by Tixxx** [comment]
@golechwierowicz @frgossen  can you guys share more info on the revert? thanks.

**Comment by golechwierowicz** [comment]
> @golechwierowicz @frgossen can you guys share more info on the revert? thanks.

The change is in after all https://github.com/openxla/xla/commit/045256bbef5ab4fea5a695efb37dd907d49c7129.

**Comment by frgossen** [comment]
The issue was an unused variable, kNumPartitions

**Pull request closed without merging**